src/services/completions.ts

```diff
@@ -658,6 +670,225 @@ function isCheckedFile(sourceFile: SourceFile, compilerOptions: CompilerOptions)
+                        break;
+                }
+            }
+        }
+        if (elements.length === 0) {
+            return undefined;
+        }
+
+        const newClauses = map(elements, element => {
+            return factory.createCaseClause(element, []);
+        });
+        const printer = createSnippetPrinter({
+            removeComments: true,
+            module: options.module,
+            target: options.target,
+            newLine: getNewLineKind(getNewLineCharacter(options, maybeBind(host, host.getNewLine))),
+        });
+        const insertText = formatContext
+            ? printer.printAndFormatSnippetList(
+                ListFormat.MultiLine | ListFormat.NoTrailingNewLine,
+                factory.createNodeArray(newClauses),
+                sourceFile,
+                formatContext)
+            : printer.printSnippetList(
+                ListFormat.MultiLine | ListFormat.NoTrailingNewLine,
+                factory.createNodeArray(newClauses),
+                sourceFile);
+
+        const firstClause = printer.printSnippetList(ListFormat.SingleLine, factory.createNodeArray([first(newClauses)!]), sourceFile);
+        return {
+            entry: {
+                name: `${firstClause} ...`,
+                kind: ScriptElementKind.unknown,
+                sortText: SortText.GlobalsOrKeywords,
+                insertText,
+                hasAction: importAdder.hasFixes() || undefined,
+                source: CompletionSource.SwitchCases,
+            },
+            importAdder,
+        };
+    }
+
+    return undefined;
+}
+
+interface CaseClauseTracker {
+    addValue(value: string | number): void;
+    hasValue(value: string | number | PseudoBigInt): boolean;
+}
+
+function newCaseClauseTracker(checker: TypeChecker, clauses: readonly (CaseClause | DefaultClause)[]): CaseClauseTracker {
+    const existingStrings = new Set<string>();
+    const existingNumbers = new Set<number>();
+    const existingBigInts = new Set<string>();
+
+    for (const clause of clauses) {
+        if (!isDefaultClause(clause)) {
+            if (isLiteralExpression(clause.expression)) {
+                const expression = clause.expression;
+                switch (expression.kind) {
+                    case SyntaxKind.NoSubstitutionTemplateLiteral:
+                    case SyntaxKind.StringLiteral:
+                        existingStrings.add(expression.text);
+                        break;
+                    case SyntaxKind.NumericLiteral:
+                        existingNumbers.add(parseInt(expression.text));
+                        break;
+                    case SyntaxKind.BigIntLiteral:
+                        const parsedBigInt = parseBigInt(endsWith(expression.text, "n") ? expression.text.slice(0, -1) : expression.text);
+                        if (parsedBigInt) {
+                            existingBigInts.add(pseudoBigIntToString(parsedBigInt));

```

Comment:
Do we need to track in some way that one of the clauses was unparseable?

Reply:
I don't think so, if the program we're completing has errors, I think (1) we'd still want to offer the completion, (2) there will probably already be a diagnostic for it. Do you have a specific thing in mind you thought we could do in such a case?